### PR TITLE
Rebase Windows support onto RawHandle instead of File

### PR DESCRIPTION
**Pull request by adamreichold**
This is a follow-up to #30 and allows mapping anything that implements `AsRawHandle` including the types of `async-std`.

I took the Win32 API usage from Rust's standard library and thereby adjusted the existing structure declarations to match.

I did test this under Wine using the `async-std` cat example as in #30 but I have to admit to not feeling particularly confident on this as Win32 is not a platform I regularly target.

### CODE REVIEW

**Comment by RazrFalcon** [comment]
Nice! Should I merge #30 first?

File: src/windows.rs [hunk 1]
```diff
@@ -156,51 +188,66 @@ impl MmapInner {
...
-                0,
-                ptr::null(),
-            );
-            if handle.is_null() {
+            let mapping = CreateFileMappingW(handle, ptr::null_mut(), protect, 0, 0, ptr::null());
+            if mapping.is_null() {
                 return Err(io::Error::last_os_error());
             }
 
+            dbg!(mapping, access, aligned_offset, aligned_len);
```
Comment (RazrFalcon) [inline src/windows.rs:204]: Remove.

File: src/windows.rs [hunk 2]
```diff
@@ -467,6 +528,16 @@ fn allocation_granularity() -> usize {
...
+    let mut info = mem::MaybeUninit::<BY_HANDLE_FILE_INFORMATION>::uninit();
+
+    unsafe {
+        let ok = GetFileInformationByHandle(handle, info.as_mut_ptr());
+        if ok == 0 {
+            return Err(io::Error::last_os_error());
+        }
+        let info = info.assume_init();
+
+        Ok((info.nFileSizeHigh as u64) << 32 | info.nFileSizeLow as u64)
```
Comment (RazrFalcon) [inline src/windows.rs:541]: Move out of `unsafe` scope. Maybe something like:

```rust
   let info = unsafe {
        let mut info = mem::MaybeUninit::<BY_HANDLE_FILE_INFORMATION>::uninit();
        if GetFileInformationByHandle(handle, info.as_mut_ptr()) == 0 {
            return Err(io::Error::last_os_error());
        }
        info.assume_init()
    }

    Ok((info.nFileSizeHigh as u64) << 32 | info.nFileSizeLow as u64)
```

**Review by RazrFalcon — CHANGES_REQUESTED** [review]

**Comment by RazrFalcon** [comment]
Adding more raw winapi calls is not great, but looks like there is no other way.

**Comment by adamreichold** [comment]
> Nice! Should I merge #30 first?

Yes, #30 is simple and safe while this is bound to get more involved.

> Adding more raw winapi calls is not great, but looks like there is no other way.

Well, I could actually use `fs::File::from_raw_handle` to turn the given handle back into an instance of `fs::File`. This way I could use basically the same API calls as now, but I am not sure if `fs::File` makes any extra assumptions on the handle (that the raw Win32 API calls do not make) and personally, I would say that roundabout way would be quite confusing.

A different maintainability/clean-up change I could follow-up with would be to replace the manual FFI declarations by usage of the `winapi` crate? But I am not sure how downstream consumers (most importantly rustc) would like that.

File: src/windows.rs (hunk 1 above)
Comment (adamreichold) [reply src/windows.rs:204]: Ooops. Let's pretend that I left this in as a proof that I did actually run and debug this. ;D

File: src/windows.rs (hunk 2 above)
Comment (adamreichold) [reply src/windows.rs]: Done as suggested.

**Comment by RazrFalcon** [comment]
The whole reason this fork exist is to remove the `winapi` crate dependency, so no. The current code if fine.

**Review by RazrFalcon — APPROVED** [review]

**Comment by RazrFalcon** [comment]
Thanks again for your work! Do you plan any other changes or can I publish a new release?

**Comment by adamreichold** [comment]
> Thanks again for your work! Do you plan any other changes or can I publish a new release?

I am currently pondering whether the assertions in the `Drop` impls for `MmapInner` should be removed in the line with the discussion at https://github.com/rust-lang/lang-team/issues/97 I am currently looking into standard library types and they all seem to ignore being unable to close/free/release resources during drop, e.g. https://doc.rust-lang.org/stable/src/std/io/buffered/bufwriter.rs.html#669-674 What do you think?

**Pull request merged**